### ISSUE TITLE
Bug: no-useless-assignment rule fails

### ISSUE THREAD

**Issue by ibc** [eslint/eslint#20513]
### Environment

Node version: 24.12.0
npm version: 11.6.2
Local ESLint version: 10.0.1
Global ESLint version:
Operating System: macOS

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```js
import eslint from '@eslint/js';
import tsEslint from 'typescript-eslint';
import jestEslint from 'eslint-plugin-jest';
import prettierRecommendedEslint from 'eslint-plugin-prettier/recommended';
import globals from 'globals';

const config = tsEslint.config(
	{
		languageOptions: {
			sourceType: 'module',
			globals: { ...globals.node },
		},
		linterOptions: {
			noInlineConfig: false,
			reportUnusedDisableDirectives: 'error',
		},
	},
	eslint.configs.recommended,
	prettierRecommendedEslint,
	{
		rules: {
			'constructor-super': 2,
			curly: [2, 'all'],
			// Unfortunatelly `curly` does not apply to blocks in `switch` cases so
			// this is needed.
			'no-restricted-syntax': [
				2,
				{
					selector: 'SwitchCase > *.consequent[type!="BlockStatement"]',
					message: 'Switch cases without blocks are disallowed',
				},
			],
			'guard-for-in': 2,
			'newline-after-var': 2,
			'newline-before-return': 2,
			'no-alert': 2,
			'no-caller': 2,
			'no-case-declarations': 2,
			'no-catch-shadow': 2,
			'no-class-assign': 2,
			'no-console': 2,
			'no-const-assign': 2,
			'no-debugger': 2,
			'no-dupe-args': 2,
			'no-dupe-keys': 2,
			'no-duplicate-case': 2,
			'no-div-regex': 2,
			'no-empty': [2, { allowEmptyCatch: true }],
			'no-empty-pattern': 2,
			'no-eval': 2,
			'no-extend-native': 2,
			'no-ex-assign': 2,
			'no-extra-bind': 2,
			'no-extra-boolean-cast': 2,
			'no-extra-label': 2,
			'no-fallthrough': 2,
			'no-func-assign': 2,
			'no-global-assign': 2,
			'no-implicit-coercion': 2,
			'no-implicit-globals': 2,
			'no-inner-declarations': 2,
			'no-invalid-regexp': 2,
			'no-invalid-this': 2,
			'no-irregular-whitespace': 2,
			'no-lonely-if': 2,
			'no-multi-str': 2,
			'no-native-reassign': 2,
			'no-negated-in-lhs': 2,
			'no-new': 2,
			'no-new-func': 2,
			'no-new-wrappers': 2,
			'no-obj-calls': 2,
			'no-proto': 2,
			'no-prototype-builtins': 0,
			'no-redeclare': 2,
			'no-regex-spaces': 2,
			'no-restricted-imports': 2,
			'no-return-assign': 2,
			'no-self-assign': 2,
			'no-self-compare': 2,
			'no-sequences': 2,
			'no-shadow': 2,
			'no-shadow-restricted-names': 2,
			'no-sparse-arrays': 2,
			'no-this-before-super': 2,
			'no-throw-literal': 2,
			'no-undef': 2,
			'no-unmodified-loop-condition': 2,
			'no-unreachable': 2,
			'no-unused-vars': [
				2,
				{ vars: 'all', args: 'after-used', caughtErrors: 'none' },
			],
			'no-use-before-define': 0,
			'no-useless-call': 2,
			'no-useless-computed-key': 2,
			'no-useless-concat': 2,
			'no-useless-rename': 2,
			'no-var': 2,
			'object-curly-newline': 0,
			'prefer-const': 2,
			'prefer-rest-params': 2,
			'prefer-spread': 2,
			'prefer-template': 2,
			'spaced-comment': [2, 'always'],
			strict: 2,
			'valid-typeof': 2,
			yoda: 2,
		},
	},
	// NOTE: We need to apply this only to .ts source files (and not to .mjs
	// files).
	...tsEslint.configs.recommendedTypeChecked.map(item => ({
		...item,
		files: ['src/**/*.ts'],
	})),
	// NOTE: We need to apply this only to .ts source files (and not to .mjs
	// files).
	...tsEslint.configs.stylisticTypeChecked.map(item => ({
		...item,
		files: ['src/**/*.ts'],
	})),
	{
		name: '.ts source files',
		files: ['src/**/*.ts'],
		languageOptions: {
			parserOptions: {
				project: 'tsconfig.json',
			},
		},
		rules: {
			'@typescript-eslint/consistent-generic-constructors': [
				2,
				'type-annotation',
			],
			'@typescript-eslint/dot-notation': 0,
			'@typescript-eslint/no-unused-vars': [
				2,
				{
					vars: 'all',
					args: 'after-used',
					caughtErrors: 'none',
					ignoreRestSiblings: false,
				},
			],
			// We want to use `type` instead of `interface`.
			'@typescript-eslint/consistent-type-definitions': 0,
			'@typescript-eslint/explicit-function-return-type': [
				2,
				{ allowExpressions: true },
			],
			// Sorry, we need many `any` usage.
			'@typescript-eslint/no-explicit-any': 0,
			'@typescript-eslint/no-empty-function': 0,
		},
	},
	{
		name: '.ts test files',
		...jestEslint.configs['flat/recommended'],
		files: ['src/test/**/*.ts'],
		rules: {
			...jestEslint.configs['flat/recommended'].rules,
			'jest/no-disabled-tests': 2,
			'jest/prefer-expect-assertions': 0,
			'@typescript-eslint/no-unnecessary-type-assertion': 0,
		},
	}
);

export default config;
```
</details>

```js
const executionsCount: Map<string, number> = new Map();

let count = executionsCount.get('foo') ?? 0;

executionsCount.set('foo', ++count);
```

### What did you expect to happen?

ESLint should not complain about anything.

### What actually happened?

ESLint fails with this error pointing to the last line in the code above:

```
error  This assigned value is not used in subsequent statements  no-useless-assignment
```

I'm declaring `let count` which gets value and then increment it in-place (`++count`). What is the ESLint rule telling to me? Should I use `const count` and then pass `count + 1`? Why?

### Link to Minimal Reproducible Example

sorry, no time for this

### Participation

- [ ] I am willing to submit a pull request for this issue.

### AI acknowledgment

- [x] I did not use AI to generate this issue report.
- [ ] (If the above is not checked) I have reviewed the AI-generated content before submitting.

**Comment by mdjermanovic** [comment]
This works as intended because the assigned value is not used in subsequent statements, as the error message says, so `++count` can be replaced with `count + 1`. While it can be argued that this is a false positive because the new value is actually used, I think it's an edge case that can easily be avoided by replacing it with `count + 1`, which I think also improves the readability of the code because it indicates the intention to use the value only in that place.
